File: src/main/java/structs/doubleLinkedList/DoublyLinkedList.java
package structs.doubleLinkedList.exercises;

public class DesignLinkedList {

    private Node head;
    private Node tail;
    private int length;

    public DesignLinkedList() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    public int get(int index) {
        if (index < 0 || index >= length) {
            return -1;
        }

        Node curr = head;
        if (index < length/2) {
            for (int i = 0; i < index; i++) {
                curr = curr.next;
            }
        } else {
            curr = tail;
            for (int i = length - 1; i > index; i--) {
                curr = curr.prev;
            }
        }

        return curr.value;
    }

    public void addAtHead(int val) {
        Node newNode = new Node(val);
        if (length == 0) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head.prev = newNode;
            head = newNode;
        }
        length++;
    }

    public void addAtTail(int val) {
        Node newNode = new Node(val);
        if (length == 0) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.prev = tail;
            tail.next = newNode;
            tail = newNode;
        }
        length++;
    }

    public void addAtIndex(int index, int val) {
        if (index == 0) {
            addAtHead(val);
        }

        if (index == length) {
            addAtTail(val);
        }

        if (index >= 0 || index < length) {
            Node newNode = new Node(val);

            Node curr = head;
            while (index < length) {
                curr = curr.next;
            }

            Node before = curr.prev;
            newNode.prev = before;
            newNode.next = curr;
            before.next = newNode;
            curr.prev = newNode;
            length++;
        }
    }

    public void deleteAtIndex(int index) {
        if (index == 0) {
            head = null;
            tail = null;
        }


    }

    public class Node {
        Node next;
        Node prev;
        int value;

        Node(int value) {
            this.value = value;
        }
    }

}

File: golang/date_structure/arrays_and_hashing/doubly_linked_list.go
package datastruc

import (
	"fmt"
)

// DoublyLinkedListNode is the node type for a doubly linked list
type DoublyLinkedListNode struct {
	val  int
	next *DoublyLinkedListNode
	prev *DoublyLinkedListNode
}

// DoublyLinkedList is the structure for a doubly linked list
type DoublyLinkedList struct {
	head *DoublyLinkedListNode
	tail *DoublyLinkedListNode
}

// NewDoublyLinkedListNode creates a new doubly linked list node
func NewDoublyLinkedListNode(val int) *DoublyLinkedListNode {
	return &DoublyLinkedListNode{
		val:  val,
		next: nil,
		prev: nil,
	}
}

// NewDoublyLinkedList creates a new doubly linked list
func NewDoublyLinkedList() *DoublyLinkedList {
	head := NewDoublyLinkedListNode(-1)
	tail := NewDoublyLinkedListNode(-1)
	head.next = tail
	tail.prev = head
	return &DoublyLinkedList{
		head: head,
		tail: tail,
	}
}

// InsertFront inserts a new node at the front of the list
func (d *DoublyLinkedList) InsertFront(val int) {
	newNode := NewDoublyLinkedListNode(val)
	newNode.prev = d.head
	newNode.next = d.head.next

	d.head.next.prev = newNode
	d.head.next = newNode
}

// InsertEnd inserts a new node at the end of the list
func (d *DoublyLinkedList) InsertEnd(val int) {
	newNode := NewDoublyLinkedListNode(val)
	newNode.next = d.tail
	newNode.prev = d.tail.prev

	d.tail.prev.next = newNode
	d.tail.prev = newNode
}

// RemoveFront removes the node at the front of the list
func (d *DoublyLinkedList) RemoveFront() {
	if d.head.next != d.tail {
		d.head.next.next.prev = d.head
		d.head.next = d.head.next.next
	}
}

// RemoveEnd removes the node at the end of the list
func (d *DoublyLinkedList) RemoveEnd() {
	if d.tail.prev != d.head {
		d.tail.prev.prev.next = d.tail
		d.tail.prev = d.tail.prev.prev
	}
}

// Print prints the doubly linked list
func (d *DoublyLinkedList) Print() {
	curr := d.head.next
	for curr != d.tail {
		fmt.Printf("%d -> ", curr.val)
		curr = curr.next
	}
	fmt.Println()
}
